extensions/amp-form/0.1/test/test-form-dirtiness.js

```diff
@@ -0,0 +1,183 @@
+      const nameless = doc.createElement('input');
+      form.appendChild(nameless);
+
+      changeInput(nameless, 'changed');
+
+      expect(form).to.not.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+
+    it('does not add dirtiness class if a hidden element changes', () => {
+      const hidden = doc.createElement('input');
+      hidden.name = 'name';
+      hidden.hidden = true;
+      form.appendChild(hidden);
+
+      changeInput(hidden, 'changed');
+
+      expect(form).to.not.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+
+    it('does not add dirtiness class if a disabled element changes', () => {
+      const disabled = doc.createElement('input');
+      disabled.name = 'name';
+      disabled.disabled = true;
+      form.appendChild(disabled);
+
+      changeInput(disabled, 'changed');
+      expect(form).to.not.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+  });
+
+  describe('text field changes', () => {
+    let textField;
+
+    beforeEach(() => {
+      // Element is inserted as HTML so that the `defaultValue` property is
+      // generated correctly, since it returns "the default value as
+      // **originally specified in the HTML** that created this object."
+      // https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement#Properties
+      const html = '<input name="name" type="text" value="default">';
+      form.insertAdjacentHTML('afterbegin', html);
+      textField = form.querySelector('input');
+    });
+
+    it('removes dirtiness class when text field is in default state', () => {
+      changeInput(textField, textField.defaultValue);
+      expect(form).to.not.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+
+    it('removes dirtiness class when text field is empty', () => {
+      changeInput(textField, '');
+      expect(form).to.not.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+
+    it('adds dirtiness class when text field is changed', () => {
+      changeInput(textField, 'changed');
+      expect(form).to.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+  });
+
+  describe('textarea changes', () => {
+    let textarea;
+
+    beforeEach(() => {
+      const html = '<textarea name="comment">default</textarea>';
+      form.insertAdjacentHTML('afterbegin', html);
+      textarea = form.querySelector('textarea');
+    });
+
+    it('removes dirtiness class when textarea is in default state', () => {
+      changeInput(textarea, textarea.defaultValue);
+      expect(form).to.not.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+
+    it('removes dirtiness class when textarea is empty', () => {
+      changeInput(textarea, '');
+      expect(form).to.not.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+
+    it('adds dirtiness class when textarea is changed', () => {
+      changeInput(textarea, 'changed');
+      expect(form).to.have.class(DIRTINESS_INDICATOR_CLASS);
+    });
+  });
+
+  describe('#onSubmitting', () => {
```

Comment:
Just curious, what does this `#` stands for, the method name?

Reply:
Ahh an odd habit. `#` denotes instance methods while `.` or `::` are for static methods. I couldn't remember/find the source of this convention tho. Just grep'ed the AMP code base -

https://github.com/ampproject/amphtml/blob/0e180322a141e9488226b80f600ded1f53823151/extensions/amp-ad-network-adsense-impl/0.1/test/test-amp-ad-network-adsense-impl.js#L120